src/Build.UnitTests/Graph/GraphLoadedFromSolution_tests.cs

```diff
@@ -0,0 +1,731 @@
+                return edgeInfos.Where(e => e.Key.Item2.Equals(node.ToConfigurationMetadata())).Select(e => e.Value);
+            }
+        }
+
+        [Fact]
+        public void GraphConstructionShouldThrowOnMissingSolutionDependencies()
+        {
+            _env.DoNotLaunchDebugger();
+
+            var solutionContents = SolutionFileBuilder.FromGraphEdges(
+                _env,
+                new Dictionary<int, int[]> {{1, null}, {2, null}},
+                new[] {("1", new[] {Guid.NewGuid().ToString("B")})}).BuildSolution();
+
+            var solutionFile = _env.CreateFile(
+                "solution.sln",
+                solutionContents)
+                .Path;
+
+            var exception = Should.Throw<InvalidProjectFileException>(
+                () =>
+                {
+                    new ProjectGraph(solutionFile);
+                });
+
+            exception.Message.ShouldContain("but a project with this GUID was not found in the .SLN file");
+        }
+
+        private static bool IsSolutionItemReference(ProjectItemInstance edgeItem)
+        {
+            return edgeItem.ItemType == GraphBuilder.SolutionItemReference;
+        }
+
+        private static bool EdgeCompliesWithSolutionDependency((ConfigurationMetadata, ConfigurationMetadata) edge, (int, int) solutionDependency)
+        {
+            return GetProjectNumber(edge.Item1) == solutionDependency.Item1 && GetProjectNumber(edge.Item2) == solutionDependency.Item2;
+        }
+
+        private void AssertSolutionBasedGraph(
+            Dictionary<int, int[]> edges,
+            SolutionConfigurationInSolution currentSolutionConfiguration,
+            IReadOnlyCollection<SolutionConfigurationInSolution> solutionConfigurations,
+            Dictionary<string, Dictionary<SolutionConfigurationInSolution, ProjectConfigurationInSolution>> projectConfigurations = null)
+        {
+            var graph = CreateProjectGraph(_env, edges);
+            var graphEdges = graph.TestOnly_Edges.TestOnly_AsConfigurationMetadata();
+
+            var solutionFileBuilder = SolutionFileBuilder.FromGraph(graph, projectConfigurations);
+
+            var solutionContents = solutionFileBuilder.BuildSolution();
+
+            var solutionPath = _env.CreateFile("TheSolution.sln", solutionContents).Path;
+            var globalProperties = currentSolutionConfiguration != null
+                ? new Dictionary<string, string>
+                {
+                    ["Configuration"] = currentSolutionConfiguration.ConfigurationName,
+                    ["Platform"] = currentSolutionConfiguration.PlatformName
+                }
+                : new Dictionary<string, string>();
+
+            var graphFromSolution = new ProjectGraph(
+                new ProjectGraphEntryPoint(
+                    solutionPath,
+                    globalProperties),
+                _env.CreateProjectCollection().Collection);
+
+            // in the solution, all nodes are entry points
```

Comment:
An interesting idea. Why isn't the solution metaproject the entry point?

Reply:
I thought about that idea but the metaproj looks horrible and is full of arbitrary accumulated garbage. Also, the graph does not work with in memory projects.
So I decided to be explicit about what behaviour to cherry-pick into the graph. I'd be fine going back to a metaproj, but then I'd rather redesign the metaproj to make it more human readable.